pkg/agent/nodeportlocal/portcache/port_table.go

```diff
@@ -64,94 +93,152 @@ func NewPortTable(start, end int) (*PortTable, error) {
 func (pt *PortTable) CleanupAllEntries() {
 	pt.tableLock.Lock()
 	defer pt.tableLock.Unlock()
-	pt.Table = make(map[int]NodePortData)
+	pt.NodePortTable = make(map[int]*NodePortData)
+	pt.PodEndpointTable = make(map[string]*NodePortData)
 }
 
-func (pt *PortTable) GetEntry(nodeport int) *NodePortData {
+func (pt *PortTable) GetEntry(ip string, port int, protocol string) *NodePortData {
 	pt.tableLock.RLock()
 	defer pt.tableLock.RUnlock()
-	data, _ := pt.Table[nodeport]
-	return &data
+	return pt.getEntryByPodIPPortProtocol(ip, port, protocol)
 }
 
 func (pt *PortTable) GetDataForPodIP(ip string) []NodePortData {
 	pt.tableLock.RLock()
 	defer pt.tableLock.RUnlock()
+	return pt.getDataForPodIP(ip)
+}
+
+func (pt *PortTable) getDataForPodIP(ip string) []NodePortData {
 	var allData []NodePortData
-	for _, data := range pt.Table {
-		if data.PodIP == ip {
-			allData = append(allData, data)
+	for i := range pt.NodePortTable {
+		if pt.NodePortTable[i].PodIP == ip {
+			allData = append(allData, *pt.NodePortTable[i])
 		}
 	}
 	return allData
 }
 
-func (pt *PortTable) GetEntryByPodIPPort(ip string, port int) *NodePortData {
-	pt.tableLock.RLock()
-	defer pt.tableLock.RUnlock()
-	return pt.getEntryByPodIPPort(ip, port)
+func (pt *PortTable) getEntryByPodIPPortProtocol(ip string, port int, protocol string) *NodePortData {
```

Comment:
this is just clunky: you do a look based on the protocol but return data which is not protocol-specific

you should lookup based on IP + Port and let the caller figure the rest out